File: python/samples/05_event_subscribe.py
# SPDX-License-Identifier: MIT
# Copyright (c) Microsoft Corporation. All rights reserved.

#!/usr/bin/env python3
"""Sample 05 — EventService subscriptions (subscribe / list / delete).

Demonstrates:
  - ctx.events.subscribe_async()  — with event_types
  - ctx.events.subscribe_async()  — with message_ids (MessageIds filter)
  - ctx.events.list_subscriptions_async()
  - ctx.events.get_subscription_async()
  - ctx.events.delete_subscription_async()
  - ctx.events.submit_test_event_async()

MessageIds subscriptions let the BMC deliver only events whose MessageId
matches one of the supplied registry-qualified IDs, e.g.:
    "Base.1.8.Success", "OpenBMC.0.1.PowerButtonPressed"

NOTE: This sample only manages subscriptions; it does not start a local
      listener.  For end-to-end event delivery see sample 06.

Usage:
    python 05_event_subscribe.py [--host HOST] [--port PORT]
"""

from __future__ import annotations

import argparse
import asyncio
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

import redfish_sdk
from redfish_sdk import AuthMode, ConnectionConfig, Credentials
from redfish_sdk.errors import RedfishConnectionError


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Sample 05 — Event Subscriptions")
    p.add_argument("--host", default="127.0.0.1")
    p.add_argument("--port", type=int, default=8000)
    p.add_argument("--user", default="admin")
    p.add_argument("--password", default="password")
    p.add_argument("--no-tls-verify", action="store_true")
    p.add_argument("--no-tls", action="store_true", help="Use plain HTTP instead of HTTPS (for simulators without SSL)")
    p.add_argument(
        "--destination",
        default="http://YOUR_LISTENER_HOST:9090/events",
        help="HTTP destination URL for the subscription",
    )
    return p.parse_args()


async def main() -> None:
    args = parse_args()

    creds = Credentials(username=args.user, password=args.password)
    config = ConnectionConfig(verify_tls=not args.no_tls_verify, use_tls=not args.no_tls)

    try:
        ctx = await redfish_sdk.connect_async(
            host=args.host,
            port=args.port,
            credentials=creds,
            auth_mode=AuthMode.SESSION,
            config=config,
        )
    except RedfishConnectionError as exc:
        print(f"[ERROR] {exc}")
        sys.exit(1)

    async with ctx:
        events = ctx.events

        sub_uris: list[str] = []

        # ── Subscription 1 — filter by EventTypes ──────────────────────
        print(f"[1] Creating EventTypes subscription → {args.destination}")
        sub_resp = await events.subscribe_async(
            destination=args.destination,
            event_types=["Alert", "ResourceUpdated", "StatusChange"],
            context="RSDK-Sample-05-EventTypes",
            protocol="Redfish",
        )
        if sub_resp.success:
            sub_uri = sub_resp.headers.get("Location") or sub_resp.body.get("@odata.id", "")
            print(f"  ✓ Subscribed — URI: {sub_uri}")
            if sub_uri:
                sub_uris.append(sub_uri)
        else:
            print(f"  ✗ Subscribe failed: HTTP {sub_resp.status_code}")

        # ── Subscription 2 — filter by MessageIds ──────────────────────
        # MessageIds is the recommended Redfish 1.5+ way to scope deliveries
        # to a precise set of registry-qualified message identifiers.
        target_message_ids = [
            "Base.1.8.Success",
            "Base.1.8.GeneralError",
            "OpenBMC.0.1.PowerButtonPressed",
        ]
        print(f"\n[2] Creating MessageIds subscription → {args.destination}")
        print(f"    MessageIds filter: {target_message_ids}")
        msgid_resp = await events.subscribe_async(
            destination=args.destination,
            message_ids=target_message_ids,
            context="RSDK-Sample-05-MessageIds",
            protocol="Redfish",
        )
        if msgid_resp.success:
            msgid_uri = msgid_resp.headers.get("Location") or msgid_resp.body.get("@odata.id", "")
            print(f"  ✓ Subscribed — URI: {msgid_uri}")
            if msgid_uri:
                sub_uris.append(msgid_uri)
        else:
            print(f"  ✗ Subscribe failed: HTTP {msgid_resp.status_code}")

        # Keep a single reference for the get/delete demo below
        sub_uri = sub_uris[0] if sub_uris else None

        # ── List subscriptions ──────────────────────────────────────────
        print("\nListing all subscriptions …")
        list_resp = await events.list_subscriptions_async()
        if list_resp.success:
            members = list_resp.body.get("Members", [])
            print(f"  Total: {len(members)}")
            for m in members:
                print(f"    {m.get('@odata.id')}")
        else:
            print(f"  ✗ List failed: HTTP {list_resp.status_code}")

        # ── Get single subscription ─────────────────────────────────────
        # Show details for each created subscription so MessageIds is visible
        for uri in sub_uris:
            print(f"\nFetching subscription detail: {uri}")
            get_resp = await events.get_subscription_async(uri)
            if get_resp.success:
                body = get_resp.body
                print(f"  Destination : {body.get('Destination')}")
                print(f"  Context     : {body.get('Context')}")
                print(f"  EventTypes  : {body.get('EventTypes')}")
                print(f"  MessageIds  : {body.get('MessageIds')}")
                print(f"  Protocol    : {body.get('Protocol')}")

        # ── Submit test event ───────────────────────────────────────────
        print("\nSubmitting test event …")
        test_resp = await events.submit_test_event_async()
        if test_resp.success:
            print("  ✓ Test event submitted")
        else:
            # Many simulators do not implement SubmitTestEvent — that is OK
            print(f"  ✗ HTTP {test_resp.status_code} (may not be supported)")

        # ── Delete all created subscriptions ───────────────────────────
        print(f"\nDeleting {len(sub_uris)} subscription(s) …")
        for uri in sub_uris:
            del_resp = await events.delete_subscription_async(uri)
            if del_resp.success:
                print(f"  ✓ Deleted: {uri}")
            else:
                print(f"  ✗ Delete failed ({uri}): HTTP {del_resp.status_code}")

        print("\n✓ Event subscription sample complete")


if __name__ == "__main__":
    asyncio.run(main())

File: rust/samples/src/bin/05_event_subscribe.rs
// SPDX-License-Identifier: MIT
// Copyright (c) Microsoft Corporation. All rights reserved.

use redfish_sdk::{connect, AuthMode, ConnectionConfig, Credentials, RedfishError};

fn parse_args() -> (String, u16, String, String, bool) {
    let mut host = "127.0.0.1".to_string();
    let mut port = 8000u16;
    let mut user = "admin".to_string();
    let mut password = "password".to_string();
    let mut no_tls = false;
    let mut it = std::env::args().skip(1);
    while let Some(a) = it.next() {
        match a.as_str() {
            "--host"          => { host = it.next().unwrap(); }
            "--port"          => { port = it.next().unwrap().parse().unwrap(); }
            "--user"          => { user = it.next().unwrap(); }
            "--password"      => { password = it.next().unwrap(); }
            "--no-tls-verify" => { no_tls = true; }
            _ => {}
        }
    }
    (host, port, user, password, no_tls)
}


#[tokio::main]
async fn main() -> Result<(), RedfishError> {
    tracing_subscriber::fmt::init();
    let (host, port, user, password, no_tls) = parse_args();
    let no_tls_proto = std::env::args().any(|a| a == "--no-tls");
    let ctx = connect(&host, port, Credentials::new(&user, &password),
        AuthMode::Session, ConnectionConfig { verify_tls: !no_tls, use_tls: !no_tls_proto, ..Default::default() }).await?;
    let events = ctx.event_service();
    let destination = "http://YOUR_LISTENER_HOST:9090/events";

    println!("Creating subscription -> {destination}");
    let sub_resp = events.subscribe(
        destination,
        vec!["Alert".into(), "ResourceUpdated".into(), "StatusChange".into()],
        vec![],
        vec![],
        vec![],
        None,
        Some("RSDK-Sample-05"),
        "Redfish",
        "RedfishEvent",
    ).await?;

    let sub_uri: Option<String> = if sub_resp.success {
        let uri = sub_resp.body.as_ref()
            .and_then(|b| b["@odata.id"].as_str()).map(str::to_string);
        println!("  Subscribed - URI: {}", uri.as_deref().unwrap_or("?"));
        uri
    } else {
        println!("  Subscribe failed: HTTP {}", sub_resp.status_code);
        None
    };

    println!("\nListing subscriptions ...");
    let lr = events.list_subscriptions().await?;
    if lr.success {
        let n = lr.body.as_ref().and_then(|b| b["Members"].as_array()).map(|a| a.len()).unwrap_or(0);
        println!("  Total: {n}");
    }

    if let Some(ref uri) = sub_uri {
        println!("\nFetching subscription: {uri}");
        let gr = events.get_subscription(uri).await?;
        if gr.success {
            if let Some(b) = &gr.body {
                println!("  Destination : {}", b["Destination"].as_str().unwrap_or("?"));
                println!("  Context     : {}", b["Context"].as_str().unwrap_or("?"));
            }
        }
    }

    println!("\nSubmitting test event ...");
    let tr = events.submit_test_event(serde_json::json!({})).await?;
    if tr.success {
        println!("  Test event submitted");
    } else {
        println!("  HTTP {} (may not be supported)", tr.status_code);
    }

    if let Some(ref uri) = sub_uri {
        let dr = events.delete_subscription(uri).await?;
        if dr.success {
            println!("  Subscription deleted");
        }
    }

    println!("\nEvent subscription sample complete");
    Ok(())
}
